extension/src/time_vector.rs

```diff
@@ -115,6 +119,76 @@ pub fn unnest<'a>(
     )
 }
 
+/// Util function to convert from *const ::std::os::raw::c_char to String
+/// TimestampTz -> *const c_char -> &CStr -> &str -> String
+pub fn timestamptz_to_string(time: pg_sys::TimestampTz) -> Result<String, Utf8Error> {
+    let char_ptr = unsafe { timestamptz_to_str(time) };
+    let c_str = unsafe { CStr::from_ptr(char_ptr) };
+    c_str.to_str().map(|s| s.to_owned())
+}
+
+#[pg_extern(immutable, schema = "toolkit_experimental", parallel_safe)]
+pub fn to_plotly<'a>(series: Timevector_TSTZ_F64<'a>) -> String {
+    format_timevector(series,"{\"times\": {{ TIMES | json_encode() | safe  }}, \"vals\": {{ VALUES | json_encode() | safe }}}".to_string())
+}
+
+#[pg_extern(immutable, schema = "toolkit_experimental", parallel_safe)]
+pub fn to_text<'a>(series: Timevector_TSTZ_F64<'a>, format_string: String) -> String {
+    format_timevector(series, format_string)
+}
+
+pub fn format_timevector<'a>(series: Timevector_TSTZ_F64<'a>, format_string: String) -> String {
+    let mut context = Context::new();
+    let mut times: Vec<String> = Vec::new();
+    let mut values: Vec<String> = Vec::new();
+    if series.has_nulls() {
+        for (i, point) in series.iter().enumerate() {
+            times.push(timestamptz_to_string(point.ts).unwrap());
+            if series.is_null_val(i) {
+                values.push("null".to_string())
+            } else {
+                match point.val.to_string().as_ref() {
+                    "NaN" | "inf" | "-inf" | "Infinity" | "-Infinity" => {
+                        panic!("All values in the series must be finite")
+                    }
+                    x => values.push(x.to_string()),
+                }
+            }
+        }
+    } else {
+        // optimized path if series does not have any nulls, but might have some NaNs/infinities
+        for point in series {
+            times.push(timestamptz_to_string(point.ts).unwrap());
+            match point.val.to_string().as_ref() {
+                "NaN" | "inf" | "-inf" | "Infinity" | "-Infinity" => {
+                    panic!("All values in the series must be finite")
+                }
+                x => values.push(x.to_string()),
+            }
+        }
+    }
+
+    // without json_encode : [[2020-01-01 00:00:00+00, 30], [2020-01-02 00:00:00+00, 45], [2020-01-03 00:00:00+00, null], [2020-01-04 00:00:00+00, 55.5], [2020-01-05 00:00:00+00, 10]]
+    // with json_encode : "[[\"2020-01-01 00:00:00+00\",\"30\"],[\"2020-01-02 00:00:00+00\",\"45\"],[\"2020-01-03 00:00:00+00\",\"null\"],[\"2020-01-04 00:00:00+00\",\"55.5\"],[\"2020-01-05 00:00:00+00\",\"10\"]]"`
+    let timevals: Vec<(String, String)> = times
+        .clone()
+        .into_iter()
+        .zip(values.clone())
+        .collect::<Vec<_>>();
+
+    context.insert("TIMES", &times);
+    context.insert("VALUES", &values);
+
+    context.insert("TIMEVALS", &timevals); // paired timevals: ex. (time,value).
```

Comment:
I'm not sure either of the representations here (line 171 and 172) are that useful, we'd really like to see this output a valid JSON vector of time, value structs.

I'm guessing we can't just specify a format string like `"{\"times\": {{ TIMES | json_encode(pretty=true) | safe  }}, \"vals\": {{ VALUES | json_encode(pretty=true) | safe }}}"` here?  Could we just parse the string ourselves and replace occurrences of `TIMEVAL` with that string?

Reply:
That makes sense, I am currently trying to get that first idea to work and if it does not I will try out the second approach.